monai/bundle/scripts.py

```diff
@@ -573,49 +572,15 @@ def get_bundle_info(
     return bundle_info[version]
 
 
-def patch_bundle_tracking(parser: ConfigParser, settings: dict) -> None:
-    """
-    Patch the loaded bundle config with a new handler logic to enable experiment tracking features.
-
-    Args:
-        parser: loaded config content to patch the handler.
-        settings: settings for the experiment tracking, should follow the pattern of default settings.
-
-    """
-    for k, v in settings["configs"].items():
-        if k in settings["handlers_id"]:
-            engine = parser.get(settings["handlers_id"][k]["id"])
-            if engine is not None:
-                handlers = parser.get(settings["handlers_id"][k]["handlers"])
-                if handlers is None:
-                    engine["train_handlers" if k == "trainer" else "val_handlers"] = [v]
-                else:
-                    handlers.append(v)
-        elif k not in parser:
-            parser[k] = v
-    # save the executed config into file
-    default_name = f"config_{time.strftime('%Y%m%d_%H%M%S')}.json"
-    filepath = parser.get("execute_config", None)
-    if filepath is None:
-        if "output_dir" not in parser:
-            # if no "output_dir" in the bundle config, default to "<bundle root>/eval"
-            parser["output_dir"] = "$@bundle_root + '/eval'"
-        # experiment management tools can refer to this config item to track the config info
-        parser["execute_config"] = parser["output_dir"] + f" + '/{default_name}'"
-        filepath = os.path.join(parser.get_parsed_content("output_dir"), default_name)
-    Path(filepath).parent.mkdir(parents=True, exist_ok=True)
-    parser.export_config_file(parser.get(), filepath)
-
-
 def run(
-    runner_id: str | Sequence[str] | None = None,
+    runner_id: str | None = None,
     meta_file: str | Sequence[str] | None = None,
     config_file: str | Sequence[str] | None = None,
     logging_file: str | None = None,
     tracking: str | dict | None = None,
     args_file: str | None = None,
     **override: Any,
-) -> list:
```

Comment:
Hi @ericspod ,

I personally hope we can force bundles to use initialize, run, and finalize to define a workflow, then other apps like MONAI FL, MONAI Label, MONAI Deploy can easily process or consume a bundle workflow with standard IDs, track in: https://github.com/Project-MONAI/model-zoo/issues/294.
That's why I deprecated the runner_id arg in the previous commit of this PR, if you think runner_id is useful to switch different running sections of a bundle config, I am OK to keep it to be backward compatible, but I don't think it's necessary to introduce new flexibility for init_id and final_id in this CLI entry.
What do you think?

Thanks.

Reply:
I'm not really liking the idea of bundles being forced into this specific pattern since a lot of people don't need these components, it's when bundles are used in specific ways that these are needed. It was the same way with Deploy where we needed to know names of specific names of components but that was only for the specific way we were planning to use bundles. I would much rather bundles stay as broad definitions that suit the most people, we're requiring enough of people to make up the definitions as it is.

Reply:
I see your point, I can add the `init_id` and `final_id`. But my concern is how can the other apps know what the `run_id`, `init_id`, and `final_id` are in a bundle config? Especially when we want to automatically launch all the bundles training & inference in some CICD environment.
And even further question: How can we know the CLI entry if bundle developer inherits from our `BundleWorkflow` and developed his own `CustomWorkflow` and put it in the `bundle/scripts/`, it's discussing in: https://github.com/Project-MONAI/model-zoo/issues/304, my initial idea is to put something in the `metadata.json`.

Thanks.

Reply:
I added `init_id` and `final_id` in the latest commit, and renamed `runner_id` to `run_id`.
To be backward compatible, I didn't change the args order in `run()`.

Thanks.

Reply:
Hi @Nic-Ma:

> how can the other apps know what the run_id, init_id, and final_id are in a bundle config?

These arguments here only pertain to the command line interface so someone invoking "run" would have to know these values for themselves, which is fine to expect. In any case if these values aren't known they should be left `None` which is what would be used in the constructor of the  `ConfigWorkflow` object anyway. Having them here gives the user on the command line the option to override them. 

> How can we know the CLI entry if bundle developer inherits from our BundleWorkflow and developed his own CustomWorkflow and put it in the bundle/scripts/?

Having the arguments here means that a bundle which has appropriate entries but doesn't provide a workflow class can still be used with initialisation and finalisation. 

> It's discussing in: https://github.com/Project-MONAI/model-zoo/issues/304, my initial idea is to define the train_workflow and infer_workflow in the metadata.json, default to our BundleWorkflow.

I'm not totally sure what the solution here  would be, we could put into the `metadata.json` the class name to instantiate and use in place of the default object, but I had always thought of this file as containing meta rather than functional data about the bundle. It's mixing descriptive and semantic information a bit but maybe I'm just being picky about it.

Reply:
I will merge this PR and let's discuss these further problems in the following PRs.

Thanks.